File: frontend/apps/ios/Pantopus/Features/Homes/Maintenance/LogMaintenanceFormViewModel.swift
//
//  LogMaintenanceFormViewModel.swift
//  Pantopus
//
//  P2.9 — Backs the Log-maintenance / Edit-maintenance form. Posts to
//  `POST /api/homes/:id/maintenance` (or PUT for an edit) and — when a
//  next-due date is provided — also creates a calendar reminder via
//  `POST /api/homes/:id/events` so the household calendar surfaces the
//  upcoming task.
//
//  Field encoding notes
//  --------------------
//  The backend `HomeMaintenanceLog` schema today only stores: task,
//  vendor, cost, recurrence, due_date, status. The richer form
//  (performed-by category, notes, photos, receipt) is captured locally
//  via `MaintenanceDraftStore` so the detail screen can render the
//  2x2 photo grid + receipt thumbnail + notes block the design calls
//  for. When the backend grows columns, the form view-model + store
//  flip in one diff.
//

// swiftlint:disable file_length

import Foundation
import Observation
import SwiftUI

/// Outbound events from the form. The hosting nav-stack listens and
/// translates these into route mutations.
public enum LogMaintenanceFormEvent: Sendable, Equatable {
    case dismiss
    case created(taskId: String)
    case updated(taskId: String)
}

/// Whether the form is creating a new task or editing an existing one.
public enum LogMaintenanceFormMode: Sendable, Equatable {
    case create
    case edit(taskId: String)
}

@Observable
@MainActor
// swiftlint:disable:next type_body_length
final class LogMaintenanceFormViewModel {
    // MARK: - Form fields

    var category: MaintenanceCategory
    var title: String
    var dateCompleted: Date
    var performedBy: MaintenancePerformedBy
    var performerName: String
    var performerContact: String
    var costText: String
    var notes: String
    var photos: [MaintenanceDraftFile]
    var receipt: MaintenanceDraftFile?
    var nextDueEnabled: Bool
    var nextDueDate: Date
    var recurrence: MaintenanceRecurrence

    // MARK: - Submit state

    private(set) var isSubmitting: Bool = false
    private(set) var submitError: String?
    private(set) var pendingEvent: LogMaintenanceFormEvent?
    /// Whether the form has been touched relative to its initial pose.
    /// Drives the close-confirm sheet in `FormShell`.
    private(set) var isDirty: Bool = false
    /// `true` while the form is fetching the existing task in edit
    /// mode. Drives a shimmer overlay on the form body.
    private(set) var isLoadingExisting: Bool = false

    // MARK: - Dependencies

    private let homeId: String
    private let mode: LogMaintenanceFormMode
    private let api: APIClient
    private let draftStore: MaintenanceDraftStore
    private let now: @Sendable () -> Date
    private var initial: Snapshot

    /// Max photos the form accepts — matches the design's "up to 4".
    static let maxPhotos: Int = 4

    /// One photo slot index → file pair. The slot index lets the UI
    /// pre-render four placeholder tiles even on a fresh form.
    struct PhotoSlot: Hashable, Identifiable {
        let id: Int
        let file: MaintenanceDraftFile?
    }

    var photoSlots: [PhotoSlot] {
        (0..<Self.maxPhotos).map { idx in
            PhotoSlot(id: idx, file: idx < photos.count ? photos[idx] : nil)
        }
    }

    /// Title is required; everything else is optional. The submit
    /// button activates when the title has at least one non-whitespace
    /// character — mirrors the backend's `task` validation.
    var canSubmit: Bool {
        !title.trimmingCharacters(in: .whitespaces).isEmpty && !isSubmitting
    }

    /// Headline for the form shell — switches between Log and Edit.
    var screenTitle: String {
        switch mode {
        case .create: "Log maintenance"
        case .edit: "Edit maintenance"
        }
    }

    var submitLabel: String {
        switch mode {
        case .create: "Log"
        case .edit: "Save"
        }
    }

    private struct Snapshot: Equatable {
        let category: MaintenanceCategory
        let title: String
        let dateCompleted: Date
        let performedBy: MaintenancePerformedBy
        let performerName: String
        let performerContact: String
        let costText: String
        let notes: String
        let photos: [MaintenanceDraftFile]
        let receipt: MaintenanceDraftFile?
        let nextDueEnabled: Bool
        let nextDueDate: Date
        let recurrence: MaintenanceRecurrence
    }

    init(
        homeId: String,
        mode: LogMaintenanceFormMode = .create,
        existing: MaintenanceTaskDTO? = nil,
        api: APIClient = .shared,
        draftStore: MaintenanceDraftStore = .shared,
        now: @escaping @Sendable () -> Date = { Date() }
    ) {
        self.homeId = homeId
        self.mode = mode
        self.api = api
        self.draftStore = draftStore
        self.now = now

        // Compute the initial pose into a Snapshot first so the
        // stored-property assignments below can read from it without
        // tripping Swift's "self used before all stored properties
        // initialized" diagnostic.
        let snapshot = Self.makeInitialSnapshot(
            mode: mode,
            existing: existing,
            draftStore: draftStore,
            baseNow: now()
        )
        category = snapshot.category
        title = snapshot.title
        dateCompleted = snapshot.dateCompleted
        performedBy = snapshot.performedBy
        performerName = snapshot.performerName
        performerContact = snapshot.performerContact
        costText = snapshot.costText
        notes = snapshot.notes
        photos = snapshot.photos
        receipt = snapshot.receipt
        nextDueEnabled = snapshot.nextDueEnabled
        nextDueDate = snapshot.nextDueDate
        recurrence = snapshot.recurrence
        initial = snapshot
    }

    private static func makeInitialSnapshot(
        mode: LogMaintenanceFormMode,
        existing: MaintenanceTaskDTO?,
        draftStore: MaintenanceDraftStore,
        baseNow: Date
    ) -> Snapshot {
        let nextWeek = baseNow.addingTimeInterval(30 * 24 * 60 * 60)
        if case let .edit(taskId) = mode {
            let dto = existing
            let stored = draftStore.draft(for: taskId)
            let inferredCategory = MaintenanceCategory.from(task: dto?.task)
            let dueParsed: Date? = {
                guard let dueIso = dto?.dueDate else { return nil }
                return MaintenanceListViewModel.parseDate(dueIso)
            }()
            return Snapshot(
                category: stored?.category ?? inferredCategory,
                title: dto?.task ?? "",
                dateCompleted: Self.parsePerformedDate(dto: dto) ?? baseNow,
                performedBy: stored?.performedBy ?? Self.inferPerformedBy(vendor: dto?.vendor),
                performerName: stored?.performerName ?? (dto?.vendor ?? ""),
                performerContact: stored?.performerContact ?? "",
                costText: Self.format(cost: dto?.cost),
                notes: stored?.notes ?? "",
                photos: stored?.photos ?? [],
                receipt: stored?.receipt,
                nextDueEnabled: dueParsed != nil,
                nextDueDate: dueParsed ?? nextWeek,
                recurrence: MaintenanceRecurrence(raw: dto?.recurrence) ?? .none
            )
        } else {
            return Snapshot(
                category: .generic,
                title: "",
                dateCompleted: baseNow,
                performedBy: .self,
                performerName: "",
                performerContact: "",
                costText: "",
                notes: "",
                photos: [],
                receipt: nil,
                nextDueEnabled: false,
                nextDueDate: nextWeek,
                recurrence: .none
            )
        }
    }

    // MARK: - Async load (edit mode)

    /// Fetches the existing task and re-applies it to the form's
    /// fields. No-op for create mode. Idempotent — safe to call from
    /// `.task { … }` and `.refreshable { … }`.
    func loadIfNeeded() async {
        guard case let .edit(taskId) = mode else { return }
        // Only refresh when we don't already have content for this id.
        if !title.isEmpty && initial.title == title { return }
        isLoadingExisting = true
        defer { isLoadingExisting = false }
        do {
            let response: GetHomeMaintenanceResponse = try await api.request(
                HomesEndpoints.maintenance(homeId: homeId)
            )
            guard let dto = response.tasks.first(where: { $0.id == taskId }) else { return }
            apply(existing: dto, taskId: taskId)
        } catch {
            // Edit mode falls back to blank fields if the load fails;
            // the user can still re-enter and save (or close).
        }
    }

    private func apply(existing dto: MaintenanceTaskDTO, taskId: String) {
        let stored = draftStore.draft(for: taskId)
        let inferredCategory = MaintenanceCategory.from(task: dto.task)
        category = stored?.category ?? inferredCategory
        title = dto.task
        if let parsed = Self.parsePerformedDate(dto: dto) {
            dateCompleted = parsed
        }
        performedBy = stored?.performedBy ?? Self.inferPerformedBy(vendor: dto.vendor)
        performerName = stored?.performerName ?? (dto.vendor ?? "")
        performerContact = stored?.performerContact ?? ""
        costText = Self.format(cost: dto.cost)
        notes = stored?.notes ?? ""
        photos = stored?.photos ?? []
        receipt = stored?.receipt
        recurrence = MaintenanceRecurrence(raw: dto.recurrence) ?? .none
        if let dueIso = dto.dueDate, let parsedDue = MaintenanceListViewModel.parseDate(dueIso) {
            nextDueEnabled = true
            nextDueDate = parsedDue
        }
        initial = currentSnapshot()
        isDirty = false
    }

    // MARK: - Mutations

    func recomputeDirty() {
        isDirty = currentSnapshot() != initial
    }

    func addPhoto(_ file: MaintenanceDraftFile) {
        guard photos.count < Self.maxPhotos else { return }
        photos.append(file)
        recomputeDirty()
    }

    func removePhoto(id: UUID) {
        photos.removeAll { $0.id == id }
        recomputeDirty()
    }

    func pickReceipt(_ file: MaintenanceDraftFile?) {
        receipt = file
        recomputeDirty()
    }

    func consumeEvent() {
        pendingEvent = nil
    }

    func cancel() {
        pendingEvent = .dismiss
    }

    // MARK: - Submit

    func submit() async {
        guard canSubmit else { return }
        isSubmitting = true
        submitError = nil
        defer { isSubmitting = false }

        let trimmedTitle = title.trimmingCharacters(in: .whitespaces)
        let vendor = vendorEncoding()
        let parsedCost = Self.parseCost(costText)
        let dueIso = nextDueEnabled ? Self.isoDayString(from: nextDueDate) : nil

        do {
            let response: HomeMaintenanceResponse
            switch mode {
            case .create:
                let req = CreateMaintenanceRequest(
                    task: trimmedTitle,
                    vendor: vendor,
                    cost: parsedCost,
                    recurrence: recurrence.rawValue,
                    dueDate: dueIso,
                    status: "completed"
                )
                response = try await api.request(
                    HomesEndpoints.createMaintenance(homeId: homeId, request: req)
                )
            case let .edit(taskId):
                let req = UpdateMaintenanceRequest(
                    task: trimmedTitle,
                    vendor: vendor,
                    cost: parsedCost,
                    recurrence: recurrence.rawValue,
                    dueDate: dueIso,
                    status: nil // Don't reset status on edit.
                )
                response = try await api.request(
                    HomesEndpoints.updateMaintenance(homeId: homeId, taskId: taskId, request: req)
                )
            }

            let taskId = response.task.id
            persistExtras(for: taskId)

            // Best-effort calendar reminder. A failure here doesn't
            // block the maintenance log — the row is already saved.
            if nextDueEnabled {
                await postCalendarReminder(taskTitle: trimmedTitle, due: nextDueDate)
            }

            Analytics.track(.ctaLogMaintenanceSubmit(result: .success))
            switch mode {
            case .create:
                pendingEvent = .created(taskId: taskId)
            case .edit:
                pendingEvent = .updated(taskId: taskId)
            }
        } catch {
            Analytics.track(.ctaLogMaintenanceSubmit(result: .error))
            submitError = (error as? APIError)?.errorDescription
                ?? "Couldn't save this maintenance entry."
        }
    }

    /// Calendar reminder for the next-due date. Reuses the existing
    /// home-events endpoint that powers the Calendar tab — no new
    /// backend wiring needed. Failures are swallowed; the toast on the
    /// calendar tab will surface the missing event on next refresh if
    /// the network call ultimately failed.
    private func postCalendarReminder(taskTitle: String, due: Date) async {
        let startAtIso = Self.isoTimestampString(from: due)
        let req = CreateHomeEventRequest(
            eventType: "maintenance",
            title: taskTitle,
            startAt: startAtIso,
            description: "Maintenance reminder",
            endAt: nil,
            locationNotes: nil,
            recurrenceRule: nil,
            assignedTo: nil,
            alertsEnabled: true
        )
        do {
            let _: HomeEventResponse = try await api.request(
                HomesEndpoints.createHomeEvent(homeId: homeId, request: req)
            )
        } catch {
            // Intentionally swallowed — calendar reminder is best-effort.
        }
    }

    private func persistExtras(for taskId: String) {
        let draft = MaintenanceDraft(
            category: category,
            performedBy: performedBy,
            performerName: performerName.trimmingCharacters(in: .whitespaces),
            performerContact: performerContact.trimmingCharacters(in: .whitespaces),
            notes: notes.trimmingCharacters(in: .whitespaces),
            photos: photos,
            receipt: receipt
        )
        draftStore.upsert(draft, for: taskId)
    }

    private func vendorEncoding() -> String? {
        let trimmed = performerName.trimmingCharacters(in: .whitespaces)
        switch performedBy {
        case .self: return nil
        case .member, .contractor:
            return trimmed.isEmpty ? nil : trimmed
        }
    }

    private func currentSnapshot() -> Snapshot {
        Snapshot(
            category: category,
            title: title,
            dateCompleted: dateCompleted,
            performedBy: performedBy,
            performerName: performerName,
            performerContact: performerContact,
            costText: costText,
            notes: notes,
            photos: photos,
            receipt: receipt,
            nextDueEnabled: nextDueEnabled,
            nextDueDate: nextDueDate,
            recurrence: recurrence
        )
    }

    // MARK: - Formatters

    static func parseCost(_ text: String) -> Decimal? {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: "$", with: "")
            .replacingOccurrences(of: ",", with: "")
        if trimmed.isEmpty { return nil }
        return Decimal(string: trimmed)
    }

    static func format(cost: Decimal?) -> String {
        guard let cost else { return "" }
        let n = NSDecimalNumber(decimal: cost)
        if n.doubleValue.truncatingRemainder(dividingBy: 1) == 0 {
            return String(format: "%.0f", n.doubleValue)
        }
        return String(format: "%.2f", n.doubleValue)
    }

    static func isoDayString(from date: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(secondsFromGMT: 0)
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }

    static func isoTimestampString(from date: Date) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f.string(from: date)
    }

    static func parsePerformedDate(dto: MaintenanceTaskDTO?) -> Date? {
        guard let iso = dto?.updatedAt ?? dto?.createdAt else { return nil }
        return MaintenanceListViewModel.parseDate(iso)
    }

    static func inferPerformedBy(vendor: String?) -> MaintenancePerformedBy {
        guard let vendor, !vendor.trimmingCharacters(in: .whitespaces).isEmpty else {
            return .self
        }
        return .contractor
    }
}

/// Repeat cadence — wraps the four backend-accepted recurrence strings
/// plus a `none` (one-time) value so the segmented control has a
/// clear "no recurrence" option.
public enum MaintenanceRecurrence: String, CaseIterable, Sendable, Hashable {
    case none = "one_time"
    case monthly
    case quarterly
    case yearly

    public init?(raw: String?) {
        guard let raw else { return nil }
        switch raw {
        case "one_time": self = .none
        case "monthly": self = .monthly
        case "quarterly": self = .quarterly
        case "yearly": self = .yearly
        default: return nil
        }
    }

    public var label: String {
        switch self {
        case .none: "None"
        case .monthly: "Monthly"
        case .quarterly: "Quarterly"
        case .yearly: "Yearly"
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/maintenance/LogMaintenanceFormViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions", "LongMethod", "LongParameterList")

package app.pantopus.android.ui.screens.homes.maintenance

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.analytics.Analytics
import app.pantopus.android.data.analytics.AnalyticsEvent
import app.pantopus.android.data.analytics.AnalyticsResult
import app.pantopus.android.data.api.models.homes.CreateHomeEventRequest
import app.pantopus.android.data.api.models.homes.CreateMaintenanceRequest
import app.pantopus.android.data.api.models.homes.MaintenanceTaskDto
import app.pantopus.android.data.api.models.homes.UpdateMaintenanceRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/** Nav-arg keys for the Log / Edit / Detail Maintenance routes. */
const val LOG_MAINTENANCE_HOME_ID_KEY = "homeId"
const val LOG_MAINTENANCE_TASK_ID_KEY = "taskId"

/** Whether the form is creating a new task or editing an existing one. */
sealed interface LogMaintenanceFormMode {
    data object Create : LogMaintenanceFormMode

    data class Edit(val taskId: String) : LogMaintenanceFormMode
}

/** Repeat cadence — wraps the backend-accepted recurrence strings plus
 *  a `None` value for the segmented control. */
enum class MaintenanceRecurrence(val raw: String, val label: String) {
    None("one_time", "None"),
    Monthly("monthly", "Monthly"),
    Quarterly("quarterly", "Quarterly"),
    Yearly("yearly", "Yearly"),
    ;

    companion object {
        fun fromRaw(raw: String?): MaintenanceRecurrence = entries.firstOrNull { it.raw == raw } ?: None
    }
}

/** Mutable form state. Equatable so the dirty-check is one comparison. */
data class LogMaintenanceFormState(
    val category: MaintenanceCategory = MaintenanceCategory.Generic,
    val title: String = "",
    val dateCompleted: Instant = Instant.now(),
    val performedBy: MaintenancePerformedBy = MaintenancePerformedBy.Self,
    val performerName: String = "",
    val performerContact: String = "",
    val costText: String = "",
    val notes: String = "",
    val photos: List<MaintenanceDraftFile> = emptyList(),
    val receipt: MaintenanceDraftFile? = null,
    val nextDueEnabled: Boolean = false,
    val nextDueDate: Instant = Instant.now().plusSeconds(SECONDS_IN_30_DAYS),
    val recurrence: MaintenanceRecurrence = MaintenanceRecurrence.None,
    val isSubmitting: Boolean = false,
    val isLoadingExisting: Boolean = false,
    val submitError: String? = null,
) {
    val canSubmit: Boolean
        get() = title.isNotBlank() && !isSubmitting

    /** Slot grid for the 2x2 photos card — pads up to four slots. */
    fun photoSlots(): List<PhotoSlot> =
        (0 until MAX_PHOTOS).map { idx ->
            PhotoSlot(index = idx, file = photos.getOrNull(idx))
        }

    /** One photo slot for rendering. */
    data class PhotoSlot(val index: Int, val file: MaintenanceDraftFile?)

    companion object {
        const val MAX_PHOTOS = 4
        private const val SECONDS_IN_30_DAYS: Long = 30L * 24 * 60 * 60
    }
}

/** Outbound event from the form VM — host listens and routes. */
sealed interface LogMaintenanceFormEvent {
    data object Dismiss : LogMaintenanceFormEvent

    data class Created(val taskId: String) : LogMaintenanceFormEvent

    data class Updated(val taskId: String) : LogMaintenanceFormEvent
}

/**
 * ViewModel backing [LogMaintenanceFormScreen]. Posts to
 * `POST /api/homes/:id/maintenance` (or PUT for an edit) and — when a
 * next-due date is provided — also creates a calendar reminder via
 * `POST /api/homes/:id/events` so the household calendar surfaces the
 * upcoming task.
 *
 * Photos / notes / receipt / performed-by are stored client-side via
 * [MaintenanceDraftStore] because the backend's `HomeMaintenanceLog`
 * schema doesn't carry those columns today. When it does, the relevant
 * fields will move onto [MaintenanceTaskDto] and this store can be
 * dropped in one diff.
 */
@HiltViewModel
class LogMaintenanceFormViewModel
    @Inject
    constructor(
        private val repo: HomesRepository,
        private val draftStore: MaintenanceDraftStore,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel() {
        private val homeId: String =
            checkNotNull(savedStateHandle[LOG_MAINTENANCE_HOME_ID_KEY]) {
                "LogMaintenanceFormViewModel requires a $LOG_MAINTENANCE_HOME_ID_KEY nav argument"
            }

        private val taskIdArg: String? = savedStateHandle[LOG_MAINTENANCE_TASK_ID_KEY]

        private val mode: LogMaintenanceFormMode =
            taskIdArg?.let { LogMaintenanceFormMode.Edit(it) } ?: LogMaintenanceFormMode.Create

        private val _form = MutableStateFlow(LogMaintenanceFormState())
        val form: StateFlow<LogMaintenanceFormState> = _form.asStateFlow()

        private val _event = MutableStateFlow<LogMaintenanceFormEvent?>(null)
        val event: StateFlow<LogMaintenanceFormEvent?> = _event.asStateFlow()

        private var initial: LogMaintenanceFormState = _form.value

        private val _isDirty = MutableStateFlow(false)
        val isDirty: StateFlow<Boolean> = _isDirty.asStateFlow()

        val screenTitle: String
            get() =
                when (mode) {
                    LogMaintenanceFormMode.Create -> "Log maintenance"
                    is LogMaintenanceFormMode.Edit -> "Edit maintenance"
                }

        val submitLabel: String
            get() =
                when (mode) {
                    LogMaintenanceFormMode.Create -> "Log"
                    is LogMaintenanceFormMode.Edit -> "Save"
                }

        fun loadIfNeeded() {
            val editTaskId = (mode as? LogMaintenanceFormMode.Edit)?.taskId ?: return
            if (_form.value.title.isNotBlank() && initial.title == _form.value.title) return
            _form.value = _form.value.copy(isLoadingExisting = true)
            viewModelScope.launch {
                when (val result = repo.getHomeMaintenance(homeId)) {
                    is NetworkResult.Success -> {
                        val task = result.data.tasks.firstOrNull { it.id == editTaskId }
                        if (task != null) applyExisting(task, editTaskId)
                        _form.value = _form.value.copy(isLoadingExisting = false)
                    }
                    is NetworkResult.Failure ->
                        _form.value = _form.value.copy(isLoadingExisting = false)
                }
            }
        }

        private fun applyExisting(
            task: MaintenanceTaskDto,
            taskId: String,
        ) {
            val stored = draftStore.draft(taskId)
            val inferredCategory = MaintenanceCategory.from(task.task)
            val performedBy =
                stored?.performedBy ?: if (task.vendor.isNullOrBlank()) {
                    MaintenancePerformedBy.Self
                } else {
                    MaintenancePerformedBy.Contractor
                }
            val parsedDue =
                task.dueDate?.let { parseDate(it) } ?: _form.value.nextDueDate
            val merged =
                LogMaintenanceFormState(
                    category = stored?.category ?: inferredCategory,
                    title = task.task,
                    dateCompleted = parseInstant(task.updatedAt ?: task.createdAt) ?: Instant.now(),
                    performedBy = performedBy,
                    performerName = stored?.performerName ?: (task.vendor.orEmpty()),
                    performerContact = stored?.performerContact.orEmpty(),
                    costText = task.cost?.let { formatCost(it) }.orEmpty(),
                    notes = stored?.notes.orEmpty(),
                    photos = stored?.photos ?: emptyList(),
                    receipt = stored?.receipt,
                    nextDueEnabled = task.dueDate != null,
                    nextDueDate = parsedDue,
                    recurrence = MaintenanceRecurrence.fromRaw(task.recurrence),
                )
            _form.value = merged
            initial = merged
            _isDirty.value = false
        }

        // MARK: - Field mutators

        fun updateCategory(value: MaintenanceCategory) = mutate { it.copy(category = value) }

        fun updateTitle(value: String) = mutate { it.copy(title = value) }

        fun updateDateCompleted(value: Instant) = mutate { it.copy(dateCompleted = value) }

        fun updatePerformedBy(value: MaintenancePerformedBy) = mutate { it.copy(performedBy = value) }

        fun updatePerformerName(value: String) = mutate { it.copy(performerName = value) }

        fun updatePerformerContact(value: String) = mutate { it.copy(performerContact = value) }

        fun updateCost(value: String) = mutate { it.copy(costText = value) }

        fun updateNotes(value: String) = mutate { it.copy(notes = value) }

        fun toggleNextDue(enabled: Boolean) = mutate { it.copy(nextDueEnabled = enabled) }

        fun updateNextDueDate(value: Instant) = mutate { it.copy(nextDueDate = value) }

        fun updateRecurrence(value: MaintenanceRecurrence) = mutate { it.copy(recurrence = value) }

        fun addPhoto(file: MaintenanceDraftFile) =
            mutate {
                if (it.photos.size >= LogMaintenanceFormState.MAX_PHOTOS) {
                    it
                } else {
                    it.copy(photos = it.photos + file)
                }
            }

        fun removePhoto(id: String) = mutate { it.copy(photos = it.photos.filterNot { p -> p.id == id }) }

        fun pickReceipt(file: MaintenanceDraftFile?) = mutate { it.copy(receipt = file) }

        fun cancel() {
            _event.value = LogMaintenanceFormEvent.Dismiss
        }

        fun consumeEvent() {
            _event.value = null
        }

        private fun mutate(transform: (LogMaintenanceFormState) -> LogMaintenanceFormState) {
            val next = transform(_form.value)
            _form.value = next
            _isDirty.value = next.copy(isSubmitting = false, submitError = null) !=
                initial.copy(isSubmitting = false, submitError = null)
        }

        // MARK: - Submit

        fun submit() {
            val current = _form.value
            if (!current.canSubmit) return
            _form.value = current.copy(isSubmitting = true, submitError = null)
            viewModelScope.launch {
                val req =
                    CreateMaintenanceRequest(
                        task = current.title.trim(),
                        vendor = encodeVendor(current),
                        cost = parseCost(current.costText),
                        recurrence = current.recurrence.raw,
                        dueDate = if (current.nextDueEnabled) formatDay(current.nextDueDate) else null,
                        status = "completed",
                    )
                val result =
                    when (val m = mode) {
                        LogMaintenanceFormMode.Create -> repo.createHomeMaintenance(homeId, req)
                        is LogMaintenanceFormMode.Edit ->
                            repo.updateHomeMaintenance(
                                homeId,
                                m.taskId,
                                UpdateMaintenanceRequest(
                                    task = req.task,
                                    vendor = req.vendor,
                                    cost = req.cost,
                                    recurrence = req.recurrence,
                                    dueDate = req.dueDate,
                                    status = null,
                                ),
                            )
                    }
                when (result) {
                    is NetworkResult.Success -> {
                        val taskId = result.data.task.id
                        persistExtras(taskId, current)
                        if (current.nextDueEnabled) {
                            postCalendarReminder(current.title.trim(), current.nextDueDate)
                        }
                        Analytics.track(AnalyticsEvent.CtaLogMaintenanceSubmit(AnalyticsResult.SUCCESS))
                        _form.value = current.copy(isSubmitting = false)
                        _event.value =
                            when (mode) {
                                LogMaintenanceFormMode.Create ->
                                    LogMaintenanceFormEvent.Created(taskId)
                                is LogMaintenanceFormMode.Edit ->
                                    LogMaintenanceFormEvent.Updated(taskId)
                            }
                    }
                    is NetworkResult.Failure -> {
                        Analytics.track(AnalyticsEvent.CtaLogMaintenanceSubmit(AnalyticsResult.ERROR))
                        _form.value =
                            current.copy(
                                isSubmitting = false,
                                submitError = result.error.message,
                            )
                    }
                }
            }
        }

        private suspend fun postCalendarReminder(
            taskTitle: String,
            due: Instant,
        ) {
            val req =
                CreateHomeEventRequest(
                    eventType = "maintenance",
                    title = taskTitle,
                    startAt = formatTimestamp(due),
                    description = "Maintenance reminder",
                    alertsEnabled = true,
                )
            // Best-effort — calendar failures don't block the maintenance row.
            repo.createHomeEvent(homeId, req)
        }

        private fun persistExtras(
            taskId: String,
            current: LogMaintenanceFormState,
        ) {
            draftStore.upsert(
                taskId,
                MaintenanceDraft(
                    category = current.category,
                    performedBy = current.performedBy,
                    performerName = current.performerName.trim(),
                    performerContact = current.performerContact.trim(),
                    notes = current.notes.trim(),
                    photos = current.photos,
                    receipt = current.receipt,
                ),
            )
        }

        private fun encodeVendor(current: LogMaintenanceFormState): String? {
            val trimmed = current.performerName.trim()
            return when (current.performedBy) {
                MaintenancePerformedBy.Self -> null
                MaintenancePerformedBy.Member,
                MaintenancePerformedBy.Contractor,
                -> trimmed.ifEmpty { null }
            }
        }

        // MARK: - Static helpers exposed for tests

        companion object {
            private val dayFormatter: DateTimeFormatter =
                DateTimeFormatter.ISO_LOCAL_DATE.withZone(ZoneOffset.UTC)
            private val timestampFormatter: DateTimeFormatter =
                DateTimeFormatter.ISO_INSTANT

            fun parseCost(text: String): BigDecimal? {
                val cleaned =
                    text
                        .trim()
                        .replace("$", "")
                        .replace(",", "")
                if (cleaned.isEmpty()) return null
                return runCatching { BigDecimal(cleaned) }.getOrNull()
            }

            fun formatCost(cost: BigDecimal): String =
                if (cost.stripTrailingZeros().scale() <= 0) {
                    cost.toBigInteger().toString()
                } else {
                    cost.toPlainString()
                }

            fun formatDay(instant: Instant): String = dayFormatter.format(instant.atZone(ZoneOffset.UTC).toLocalDate())

            fun formatTimestamp(instant: Instant): String = timestampFormatter.format(instant)

            fun parseDate(iso: String): Instant {
                // Accept ISO-8601 timestamps and bare YYYY-MM-DD strings.
                return runCatching { Instant.parse(iso) }
                    .recoverCatching {
                        LocalDate.parse(iso).atStartOfDay(ZoneId.of("UTC")).toInstant()
                    }
                    .getOrElse { Instant.now() }
            }

            fun parseInstant(iso: String?): Instant? {
                if (iso.isNullOrBlank()) return null
                return runCatching { Instant.parse(iso) }
                    .recoverCatching {
                        LocalDate.parse(iso).atStartOfDay(ZoneId.of("UTC")).toInstant()
                    }
                    .getOrNull()
            }
        }
    }
